sql/core/src/test/scala/org/apache/spark/sql/connector/WriteDistributionAndOrderingSuite.scala

```diff
@@ -685,54 +788,88 @@ class WriteDistributionAndOrderingSuite
-    }
+      if (expectAnalysisException) {
+        intercept[AnalysisException] {
+          executeCommand()
+        }
+      } else if (!testSkewed) {
+        withSQLConf(
+          SQLConf.COALESCE_PARTITIONS_ENABLED.key -> "false") {
+          val executedPlan = executeCommand()
+
+          checkPartitioningAndOrdering(
+            executedPlan, expectedWritePartitioning, expectedWriteOrdering)
+
+          checkAnswer(spark.table(tableNameAsString), df)
+        }
+      } else {
+        withSQLConf(
+          SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true",
+          SQLConf.COALESCE_PARTITIONS_ENABLED.key -> "true",
+          SQLConf.ADAPTIVE_OPTIMIZE_SKEWS_IN_REBALANCE_PARTITIONS_ENABLED.key -> "true",
+          SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1",
+          SQLConf.SHUFFLE_PARTITIONS.key -> "5",
+          SQLConf.ADVISORY_PARTITION_SIZE_IN_BYTES.key -> "100",
+          SQLConf.COALESCE_PARTITIONS_MIN_PARTITION_NUM.key -> "1") {
+          val executedPlan = executeCommand()
+
+          checkPartitioningAndOrdering(
+            executedPlan, expectedWritePartitioning, expectedWriteOrdering, true)
+
+          checkAnswer(spark.table(tableNameAsString), df)
+        }
+     }
+    catalog.dropTable(ident)
   }
 
   private def checkPartitioningAndOrdering(
       plan: SparkPlan,
       partitioning: physical.Partitioning,
-      ordering: Seq[catalyst.expressions.SortOrder]): Unit = {
+      ordering: Seq[catalyst.expressions.SortOrder],
+      testSkewed: Boolean = false): Unit = {
 
     val sorts = collect(plan) { case s: SortExec => s }
     assert(sorts.size <= 1, "must be at most one sort")
     val shuffles = collect(plan) { case s: ShuffleExchangeLike => s }
     assert(shuffles.size <= 1, "must be at most one shuffle")
 
     val actualPartitioning = plan.outputPartitioning
-    val expectedPartitioning = partitioning match {
-      case p: physical.RangePartitioning =>
-        val resolvedOrdering = p.ordering.map(resolveAttrs(_, plan))
-        p.copy(ordering = resolvedOrdering.asInstanceOf[Seq[catalyst.expressions.SortOrder]])
-      case p: physical.HashPartitioning =>
-        val resolvedExprs = p.expressions.map(resolveAttrs(_, plan))
-        p.copy(expressions = resolvedExprs)
-      case other => other
+    if (!testSkewed) {
+      val expectedPartitioning = partitioning match {
+        case p: physical.RangePartitioning =>
+          val resolvedOrdering = p.ordering.map(resolveAttrs(_, plan))
+          p.copy(ordering = resolvedOrdering.asInstanceOf[Seq[catalyst.expressions.SortOrder]])
+        case p: physical.HashPartitioning =>
+          val resolvedExprs = p.expressions.map(resolveAttrs(_, plan))
+          p.copy(expressions = resolvedExprs)
+        case other => other
+      }
+      assert(actualPartitioning == expectedPartitioning, "partitioning must match")
+    } else {
+      assert(actualPartitioning.numPartitions > conf.numShufflePartitions)
```

Comment:
can we collect the `AQEShuffleReadExec` in the plan and make sure we do split the skewed partitions?